Add writers for xdr and ascii files

trossi opened the pull request:
## References to issues or other PRs

Closes #31.

## Describe the proposed changes

This PR will add basic writing functionality.

Demo (`demo.py`):
```python
import numpy as np
from rdata.io import write
from rdata.conversion import convert_to_r_data

matrix = np.pi * np.arange(12.).reshape(3, 4)
data = [matrix, 'hello', dict(my_key='my_value'), None, 1, 2.2, 3.3+4.4j]
print(data)

r_data = convert_to_r_data(data)
write("demo.rds", r_data)
```

Output:
```
$ python3 demo.py 
[array([[ 0.        ,  3.14159265,  6.28318531,  9.42477796],
       [12.56637061, 15.70796327, 18.84955592, 21.99114858],
       [25.13274123, 28.27433388, 31.41592654, 34.55751919]]), 'hello', {'my_key': 'my_value'}, None, 1, 2.2, (3.3+4.4j)]

$ Rscript -e "readRDS('demo.rds')"
[[1]]
         [,1]      [,2]      [,3]      [,4]
[1,]  0.00000  3.141593  6.283185  9.424778
[2,] 12.56637 15.707963 18.849556 21.991149
[3,] 25.13274 28.274334 31.415927 34.557519

[[2]]
[1] "hello"

[[3]]
[[3]]$my_key
[1] "my_value"


[[4]]
NULL

[[5]]
[1] 1

[[6]]
[1] 2.2

[[7]]
[1] 3.3+4.4i
```

## Additional information

### Testing

Two tests looping over all test files in the repository have been added:
1. `rdata/tests/test_write.py::test_write`: Test that read-parse-write results in bit-wise unchanged files (omitting compression)
2. `rdata/tests/test_write.py::test_convert_to_r`: Test that R-to-Python-to-R conversion results in unchanged RData object.

Implemented features are indicated in the test output: `XFAIL`: writing/converting of the given type has not implemented in this PR, `SKIPPED`: R-to-Python-to-R conversion might be ambiguous (not enough information in Python object to construct the original R object).

<details>
<summary>Test output (click to expand)</summary>

```
rdata/tests/test_write.py::test_write[test_altrep_compact_intseq.rda] XFAIL (RObjectType.ALTREP)
rdata/tests/test_write.py::test_write[test_altrep_compact_intseq_asymmetric.rda] XFAIL (RObjectType.ALTREP)
rdata/tests/test_write.py::test_write[test_altrep_compact_realseq.rda] XFAIL (RObjectType.ALTREP)
rdata/tests/test_write.py::test_write[test_altrep_compact_realseq_asymmetric.rda] XFAIL (RObjectType.ALTREP)
rdata/tests/test_write.py::test_write[test_altrep_deferred_string.rda] XFAIL (RObjectType.ALTREP)
rdata/tests/test_write.py::test_write[test_altrep_wrap_logical.rda] XFAIL (RObjectType.ALTREP)
rdata/tests/test_write.py::test_write[test_altrep_wrap_real.rda] XFAIL (RObjectType.ALTREP)
rdata/tests/test_write.py::test_write[test_altrep_wrap_string.rda] XFAIL (RObjectType.ALTREP)
rdata/tests/test_write.py::test_write[test_ascii_v2.rda] PASSED
rdata/tests/test_write.py::test_write[test_ascii_v2.rds] PASSED
rdata/tests/test_write.py::test_write[test_ascii_v3.rda] PASSED
rdata/tests/test_write.py::test_write[test_ascii_v3.rds] PASSED
rdata/tests/test_write.py::test_write[test_ascii_win_v2.rda] PASSED
rdata/tests/test_write.py::test_write[test_ascii_win_v2.rds] PASSED
rdata/tests/test_write.py::test_write[test_ascii_win_v3.rda] PASSED
rdata/tests/test_write.py::test_write[test_ascii_win_v3.rds] PASSED
rdata/tests/test_write.py::test_write[test_builtin.rda] PASSED
rdata/tests/test_write.py::test_write[test_complex.rda] PASSED
rdata/tests/test_write.py::test_write[test_dataframe.rda] XFAIL (RObjectType.REF)
rdata/tests/test_write.py::test_write[test_dataframe.rds] XFAIL (RObjectType.REF)
rdata/tests/test_write.py::test_write[test_dataframe_rownames.rda] XFAIL (RObjectType.REF)
rdata/tests/test_write.py::test_write[test_dataframe_v3.rda] XFAIL (RObjectType.REF)
rdata/tests/test_write.py::test_write[test_dataframe_v3.rds] XFAIL (RObjectType.REF)
rdata/tests/test_write.py::test_write[test_empty_function.rda] XFAIL (RObjectType.CLO)
rdata/tests/test_write.py::test_write[test_empty_function_uncompiled.rda] XFAIL (RObjectType.CLO)
rdata/tests/test_write.py::test_write[test_empty_str.rda] PASSED
rdata/tests/test_write.py::test_write[test_emptyenv.rda] XFAIL (RObjectType.EMPTYENV)
rdata/tests/test_write.py::test_write[test_encodings.rda] PASSED
rdata/tests/test_write.py::test_write[test_encodings_v3.rda] PASSED
rdata/tests/test_write.py::test_write[test_environment.rda] XFAIL (RObjectType.ENV)
rdata/tests/test_write.py::test_write[test_expression.rda] PASSED
rdata/tests/test_write.py::test_write[test_file.rda] XFAIL (RObjectType.EXTPTR)
rdata/tests/test_write.py::test_write[test_full_named_matrix.rda] PASSED
rdata/tests/test_write.py::test_write[test_full_named_matrix.rds] PASSED
rdata/tests/test_write.py::test_write[test_function.rda] XFAIL (RObjectType.CLO)
rdata/tests/test_write.py::test_write[test_function_arg.rda] XFAIL (RObjectType.CLO)
rdata/tests/test_write.py::test_write[test_half_named_matrix.rda] PASSED
rdata/tests/test_write.py::test_write[test_list.rda] PASSED
rdata/tests/test_write.py::test_write[test_list_attrs.rda] PASSED
rdata/tests/test_write.py::test_write[test_logical.rda] PASSED
rdata/tests/test_write.py::test_write[test_matrix.rda] PASSED
rdata/tests/test_write.py::test_write[test_minimal_function.rda] XFAIL (RObjectType.CLO)
rdata/tests/test_write.py::test_write[test_minimal_function_uncompiled.rda] XFAIL (RObjectType.CLO)
rdata/tests/test_write.py::test_write[test_na_string.rda] PASSED
rdata/tests/test_write.py::test_write[test_named_matrix.rda] PASSED
rdata/tests/test_write.py::test_write[test_nullable_int.rda] PASSED
rdata/tests/test_write.py::test_write[test_nullable_logical.rda] PASSED
rdata/tests/test_write.py::test_write[test_s4.rda] XFAIL (RObjectType.S4)
rdata/tests/test_write.py::test_write[test_ts.rda] PASSED
rdata/tests/test_write.py::test_write[test_vector.rda] PASSED
rdata/tests/test_write.py::test_convert_to_r[test_altrep_compact_intseq.rda] SKIPPED (Type RObjectType.ALTREP not implemented)
rdata/tests/test_write.py::test_convert_to_r[test_altrep_compact_intseq_asymmetric.rda] SKIPPED (Type RObjectType.ALTREP not implemented)
rdata/tests/test_write.py::test_convert_to_r[test_altrep_compact_realseq.rda] SKIPPED (Type RObjectType.ALTREP not implemented)
rdata/tests/test_write.py::test_convert_to_r[test_altrep_compact_realseq_asymmetric.rda] SKIPPED (Type RObjectType.ALTREP not implemented)
rdata/tests/test_write.py::test_convert_to_r[test_altrep_deferred_string.rda] SKIPPED (Type RObjectType.ALTREP not implemented)
rdata/tests/test_write.py::test_convert_to_r[test_altrep_wrap_logical.rda] SKIPPED (Type RObjectType.ALTREP not implemented)
rdata/tests/test_write.py::test_convert_to_r[test_altrep_wrap_real.rda] SKIPPED (Type RObjectType.ALTREP not implemented)
rdata/tests/test_write.py::test_convert_to_r[test_altrep_wrap_string.rda] SKIPPED (Type RObjectType.ALTREP not implemented)
rdata/tests/test_write.py::test_convert_to_r[test_ascii_v2.rda] PASSED
rdata/tests/test_write.py::test_convert_to_r[test_ascii_v2.rds] PASSED
rdata/tests/test_write.py::test_convert_to_r[test_ascii_v3.rda] PASSED
rdata/tests/test_write.py::test_convert_to_r[test_ascii_v3.rds] PASSED
rdata/tests/test_write.py::test_convert_to_r[test_ascii_win_v2.rda] PASSED
rdata/tests/test_write.py::test_convert_to_r[test_ascii_win_v2.rds] PASSED
rdata/tests/test_write.py::test_convert_to_r[test_ascii_win_v3.rda] PASSED
rdata/tests/test_write.py::test_convert_to_r[test_ascii_win_v3.rds] PASSED
rdata/tests/test_write.py::test_convert_to_r[test_builtin.rda] XFAIL (<class 'rdata.conversion._conversion.RBuiltin'>)
rdata/tests/test_write.py::test_convert_to_r[test_complex.rda] PASSED
rdata/tests/test_write.py::test_convert_to_r[test_dataframe.rda] XFAIL (<class 'pandas.core.frame.DataFrame'>)
rdata/tests/test_write.py::test_convert_to_r[test_dataframe.rds] XFAIL (<class 'pandas.core.frame.DataFrame'>)
rdata/tests/test_write.py::test_convert_to_r[test_dataframe_rownames.rda] XFAIL (<class 'pandas.core.frame.DataFrame'>)
rdata/tests/test_write.py::test_convert_to_r[test_dataframe_v3.rda] XFAIL (<class 'pandas.core.frame.DataFrame'>)
rdata/tests/test_write.py::test_convert_to_r[test_dataframe_v3.rds] XFAIL (<class 'pandas.core.frame.DataFrame'>)
rdata/tests/test_write.py::test_convert_to_r[test_empty_function.rda] XFAIL (<class 'rdata.conversion._conversion.RFunction'>)
rdata/tests/test_write.py::test_convert_to_r[test_empty_function_uncompiled.rda] XFAIL (<class 'rdata.conversion._conversion.RFunction'>)
rdata/tests/test_write.py::test_convert_to_r[test_empty_str.rda] PASSED
rdata/tests/test_write.py::test_convert_to_r[test_emptyenv.rda] XFAIL (<class 'rdata.conversion._conversion.REnvironment'>)
rdata/tests/test_write.py::test_convert_to_r[test_encodings.rda] SKIPPED (ambiguous R->py->R transformation)
rdata/tests/test_write.py::test_convert_to_r[test_encodings_v3.rda] SKIPPED (ambiguous R->py->R transformation)
rdata/tests/test_write.py::test_convert_to_r[test_environment.rda] XFAIL (<class 'rdata.conversion._conversion.REnvironment'>)
rdata/tests/test_write.py::test_convert_to_r[test_expression.rda] PASSED
rdata/tests/test_write.py::test_convert_to_r[test_file.rda] SKIPPED (ambiguous R->py->R transformation)
rdata/tests/test_write.py::test_convert_to_r[test_full_named_matrix.rda] XFAIL (<class 'xarray.core.dataarray.DataArray'>)
rdata/tests/test_write.py::test_convert_to_r[test_full_named_matrix.rds] XFAIL (<class 'xarray.core.dataarray.DataArray'>)
rdata/tests/test_write.py::test_convert_to_r[test_function.rda] XFAIL (<class 'rdata.conversion._conversion.RFunction'>)
rdata/tests/test_write.py::test_convert_to_r[test_function_arg.rda] XFAIL (<class 'rdata.conversion._conversion.RFunction'>)
rdata/tests/test_write.py::test_convert_to_r[test_half_named_matrix.rda] XFAIL (<class 'xarray.core.dataarray.DataArray'>)
rdata/tests/test_write.py::test_convert_to_r[test_list.rda] PASSED
rdata/tests/test_write.py::test_convert_to_r[test_list_attrs.rda] SKIPPED (ambiguous R->py->R transformation)
rdata/tests/test_write.py::test_convert_to_r[test_logical.rda] PASSED
rdata/tests/test_write.py::test_convert_to_r[test_matrix.rda] PASSED
rdata/tests/test_write.py::test_convert_to_r[test_minimal_function.rda] XFAIL (<class 'rdata.conversion._conversion.RFunction'>)
rdata/tests/test_write.py::test_convert_to_r[test_minimal_function_uncompiled.rda] XFAIL (<class 'rdata.conversion._conversion.RFunction'>)
rdata/tests/test_write.py::test_convert_to_r[test_na_string.rda] PASSED
rdata/tests/test_write.py::test_convert_to_r[test_named_matrix.rda] XFAIL (<class 'xarray.core.dataarray.DataArray'>)
rdata/tests/test_write.py::test_convert_to_r[test_nullable_int.rda] PASSED
rdata/tests/test_write.py::test_convert_to_r[test_nullable_logical.rda] PASSED
rdata/tests/test_write.py::test_convert_to_r[test_s4.rda] XFAIL (<class 'types.SimpleNamespace'>)
rdata/tests/test_write.py::test_convert_to_r[test_ts.rda] XFAIL (<class 'pandas.core.series.Series'>)
rdata/tests/test_write.py::test_convert_to_r[test_vector.rda] PASSED
```

</details>


These tests do not cover all the functionality of writer, so extra tests need to be added still.

### Documentation

No documentation in README or https://rdata.readthedocs.io is included in this PR yet.


## Checklist before requesting a review

- [x] I have performed a self-review of my code
- [x] The code conforms to the style used in this package (checked with [Ruff](https://docs.astral.sh/ruff/))
- [x] The code is fully documented and typed (type-checked with [Mypy](https://mypy-lang.org/))
- [x] I have added thorough tests for the new/changed functionality

trossi commented:
Hi @vnmabus! Sorry for the delay with this PR. Could you have a look what you think in general about this PR and if such a writing feature would fit within the rdata package?

There is still some work to be done, for example some usage documentation and more tests would still be needed, and the organization of files and the names of classes could be streamlined (e.g., should "Writer" be called something else to match better as the counterpart of "Parser"?).

vnmabus commented on pyproject.toml:
```diff
@@ -127,6 +127,9 @@ ignore = [
 	"S310", # URLs in examples have been validated
 	"T201", # print allowed in examples
 ]
+"rdata/tests/*.py" = [
```
I do not think tests should print. Using assertions or raising exceptions seems better IMHO.

trossi replied:
Removed prints c0a118ec858ae9f0357b8535b8c67f715912d0ee. The reason for adding these was to get a more readable output for the cases when the assert fails.

vnmabus commented on rdata/conversion/__init__.py:
```diff
@@ -24,3 +24,7 @@
     factor_constructor as factor_constructor,
     ts_constructor as ts_constructor,
 )
+from .to_r import (
```
We probably need a better, more symmetrical name for this functionality in the future, but maybe we should left that to a future refactor.

trossi replied:
I agree. (Commented more below.)

vnmabus commented on rdata/tests/test_write.py:
```diff
@@ -0,0 +1,114 @@
+        out_data = fd.getvalue()
+
+        if data != out_data:
+            print(r_data)
```
Remove prints in tests.

trossi replied:
Done c0a118ec858ae9f0357b8535b8c67f715912d0ee.

vnmabus reviewed (changes requested):
print statements should be removed from tests.

I think that naming and file organization could be improved, but probably is better to do that after the merge. I see no other obvious problems.

I think that this functionality is good for achieving limited intercompatibility (there was even a suggestion in #20), although users should be aware that information loss may happen when converting forth and back.

trossi commented:
Thank you for the review! I did the changes and pushed some extra tests too.

> I think that naming and file organization could be improved, but probably is better to do that after the merge. I see no other obvious problems.

I agree that the current naming does not align with the existing names. I left the naming open for you to decide what should be the suitable naming convention (e.g. should the current "Writer" be "Deparser" or something else instead to align with "Parser"). I can do some renaming too if you prefer.

But it would be great if we could have already in this PR somewhat stable high-level user interface. For example, is it ok for you if I add `write_rds` etc (analogously to `read_rds` etc) to `rdata/__init__.py`, so that a script like below would keep working even when internal names are changed (similar to the demo code in PR description)?

```python
import numpy as np
from rdata import write_rds

matrix = np.pi * np.arange(12.).reshape(3, 4)
data = [matrix, 'hello', dict(my_key='my_value'), None, 1, 2.2, 3.3+4.4j]

write_rds("demo.rds", data)
```

Regarding our earlier discussion about authors in pyproject.toml, would you mind including me there? I feel that implementing this initial writer support is somewhat sizeable contribution to the package.

trossi commented:
Is there a way to regenerate the code coverage analysis? The coverage should be now better than reported above, but I could still add more tests to cover missing branches.

vnmabus commented:
> I agree that the current naming does not align with the existing names. I left the naming open for you to decide what should be the suitable naming convention (e.g. should the current "Writer" be "Deparser" or something else instead to align with "Parser"). I can do some renaming too if you prefer.
> 
> But it would be great if we could have already in this PR somewhat stable high-level user interface. For example, is it ok for you if I add `write_rds` etc (analogously to `read_rds` etc) to `rdata/__init__.py`, so that a script like below would keep working even when internal names are changed (similar to the demo code in PR description)?

`write_rds` and `write_rdata` are clear candidates for a stable high-level user interface.

For the parsing step, we could use `unparse_file` and `unparse_data`, as mirror images of `parse_file` and `parse_data`. The `unparse` nomenclature was used [in the `ast` module](https://docs.python.org/3/library/ast.html#ast.unparse), so I hope it is not too surprising.

The `convert` analog is less clear, as the name `convert` does not imply a particular direction. I would propose to add two methods to `Converter`, as well as the corresponding functions: `python_to_r` and `r_to_python`. `convert` would then be an alias to `r_to_python`, and we may decide to deprecate it.

> Regarding our earlier discussion about authors in pyproject.toml, would you mind including me there? I feel that implementing this initial writer support is somewhat sizeable contribution to the package.

I do not mind at all. There is no particular contribution threshold to appear there. Please add yourself to pyproject.toml, either as part of this PR or in a different one, and I will review it when I have time (hopefully soon). Is there a particular reason for you to want to appear there?

trossi commented:
> `write_rds` and `write_rdata` are clear candidates for a stable high-level user interface.

I added this interface to `rdata/_write.py` and `rdata/__init__.py`.
 
> For the parsing step, we could use `unparse_file` and `unparse_data`, as mirror images of `parse_file` and `parse_data`. The `unparse` nomenclature was used [in the `ast` module](https://docs.python.org/3/library/ast.html#ast.unparse), so I hope it is not too surprising.

I have renamed writing to unparsing throughout to have more symmetric naming with parsers and added also `unparse_data()` operating on bytestring similarly to `parse_data()`.

> The `convert` analog is less clear, as the name `convert` does not imply a particular direction. I would propose to add two methods to `Converter`, as well as the corresponding functions: `python_to_r` and `r_to_python`. `convert` would then be an alias to `r_to_python`, and we may decide to deprecate it.

I feel that having both conversions under a single `Converter` class might make it rather complex. For example, some state variables would not likely be needed for both directions (e.g., the current R-to-Python `Converter` has `default_encoding` and `force_default_encoding` for handling various encodings the input RData object might contain, but Python-to-R is simpler in this respect as we need to only specify which encoding to use for the output RData).

I didn't do any refactoring for the conversion part and it is as before.

> I do not mind at all. There is no particular contribution threshold to appear there. Please add yourself to pyproject.toml, either as part of this PR or in a different one, and I will review it when I have time (hopefully soon). Is there a particular reason for you to want to appear there?

Great, thanks, added! I'm thinking this in case it might be useful when reporting work (not sure if it makes any difference though). I left emails out from the authors list so that you remain as the main contact point (in the maintainers list too with an email).

I also added a few more tests and the following section to `pyproject.toml` to exclude type checking sections from coverage report (3d815f1):

```
[tool.coverage.report]
exclude_also = [
    "if TYPE_CHECKING:",
    ]
```

Let me know if this change is ok for you.

@vnmabus This PR would be ready from my side now unless you have some change requests.

vnmabus commented:
I will review as soon as I can.

> I feel that having both conversions under a single Converter class might make it rather complex. For example, some state variables would not likely be needed for both directions (e.g., the current R-to-Python Converter has default_encoding and force_default_encoding for handling various encodings the input RData object might contain, but Python-to-R is simpler in this respect as we need to only specify which encoding to use for the output RData).

I am not heavily opposed to have two kinds of converters either, although the advantage of having one is that we could try to make roundtrip (R -> Python -> R and Python -> R -> Python) work for at least the common cases. If we use different classes for each conversion it is not immediately obvious for which combinations rountrip even makes sense.

> Great, thanks, added! I'm thinking this in case it might be useful when reporting work (not sure if it makes any difference though). I left emails out from the authors list so that you remain as the main contact point (in the maintainers list too with an email).

I asked because one of the things I wanted to do when I am less busy is to submit the package to the [Journal of Open Source Software (JOSS)](https://joss.theoj.org/). As the package was already approved by PyOpenSci, it should be fast-tracked (provided that they consider it in scope), so I would only need to write a small paper. I already have a description of the library from my PhD thesis, so I would only need to adapt it to the final format. If we merge this before submitting to JOSS, I would need to also add a small description about the writing capabilities, and I could add you as coauthor, if you need/want to be recognized. Note that although the JOSS is a peer-reviewed journal, it is still not indexed in Web of Science, so it may not be taken into account for academic merits in some countries, such as Spain, which tend to disregard publications not in JCR.

> I also added a few more tests and the following section to pyproject.toml to exclude type checking sections from coverage report (https://github.com/vnmabus/rdata/commit/3d815f1b8c8fdcd672a8e4b4bed40f9d2a746379).

That looks useful indeed. I would have done it in a separate PR, but it does not matter much.

vnmabus commented on rdata/_write.py:
```diff
@@ -0,0 +1,132 @@
+    path: os.PathLike[Any] | str,
+    data: Any,  # noqa: ANN401
+    *,
+    file_format: str = "xdr",
```
Shouldn't we use `Literal` for these strings, as the number of options is limited?

trossi replied:
I agree, good point. I changed to `Literal` and removed this duplicated information from docstring.

It might be even better to use `RdataFormats` enum from parser here. I didn't do that as the enum would need a bit of extra code to allow convenient use like `rdata.write_rds(..., file_format="xdr")` instead of heavier `rdata.write_rds(..., file_format=RdataFormats.XDR)`. This might be something to refactor later? (At the same time, it might be good to move `RdataFormats` to `rdata` from `rdata.parser` as it'd be used for both parsing and unparsing.) The same applies to  `rds` and `compression` parameters (`FileTypes` enum in parser).

vnmabus replied:
We could make it accept both, e.g. `Literal["xdr", "ascii"] | RdataFormats`. We can leave that to a later refactor.

vnmabus commented on rdata/_write.py:
```diff
@@ -0,0 +1,132 @@
+    *,
+    file_format: str = "xdr",
+    rds: bool = True,
+    compression: str = "gzip",
```
I would also use `Literal` for this option. In addition to strings, this should also accept `None`.

trossi replied:
Done.

vnmabus commented on rdata/_write.py:
```diff
@@ -0,0 +1,132 @@
+    data: Any,  # noqa: ANN401
+    *,
+    file_format: str = "xdr",
+    rds: bool = True,
```
I am not sure that this should be a boolean. Why not an enum or a `Literal` with two string values?

trossi replied:
I changed it to `file_type : Literal["rds", "rda"]`.

Similar rds flags were also in conversion and Unparser because rda files need special handling. These are now removed too to simplify and there is following instead: 
* rda-specific `convert_to_r_object_for_rda()` added (called in `rdata.write_rda()`)
* rda-specific magic bytes (duplicating information about version and xdr vs ascii) are written in `unparse_fileobj()` instead of Unparser's method

vnmabus commented on rdata/_write.py:
```diff
@@ -0,0 +1,132 @@
+    rds: bool = True,
+    compression: str = "gzip",
+    encoding: str = "UTF-8",
+    versions: tuple[int, int, int] | None = None,
```
I am not convinced this should be a tuple, instead of separate version arguments. What do you think?

trossi replied:
I agree. I simplified it even further as only the format version (2 vs 3) should be relevant for users to choose in this high-level function.

vnmabus replied:
I totally agree with that.

trossi commented:
> I am not heavily opposed to have two kinds of converters either, although the advantage of having one is that we could try to make roundtrip (R -> Python -> R and Python -> R -> Python) work for at least the common cases. If we use different classes for each conversion it is not immediately obvious for which combinations rountrip even makes sense.

Such roundtrip is currently implemented in the added test, but it's true that it's not obvious without running the code which types work for the roundtrip. This could indeed be refactored.


> I asked because one of the things I wanted to do when I am less busy is to submit the package to the [Journal of Open Source Software (JOSS)](https://joss.theoj.org/). As the package was already approved by PyOpenSci, it should be fast-tracked (provided that they consider it in scope), so I would only need to write a small paper. I already have a description of the library from my PhD thesis, so I would only need to adapt it to the final format. If we merge this before submitting to JOSS, I would need to also add a small description about the writing capabilities, and I could add you as coauthor, if you need/want to be recognized. Note that although the JOSS is a peer-reviewed journal, it is still not indexed in Web of Science, so it may not be taken into account for academic merits in some countries, such as Spain, which tend to disregard publications not in JCR.

I think JOSS is a great journal and I would be happy to co-author if that is fine with you. Thanks a lot for asking! I can also contribute in writing the paper and adding documentation about the writing functionality to the readthedocs pages. But it's up to you if you would prefer to restrict the JOSS paper to the reading capabilities of the package only.

vnmabus commented on rdata/conversion/to_r.py:
```diff
@@ -0,0 +1,330 @@
+if TYPE_CHECKING:
+    from collections.abc import Mapping
+
+    from mypy_extensions import NamedArg
```
Don't use non-standard constructs for this. Just define a [Callback protocol](https://mypy.readthedocs.io/en/stable/protocols.html#callback-protocols).

trossi replied:
Fixed, that's indeed clearer. Thanks!

vnmabus commented on rdata/conversion/to_r.py:
```diff
@@ -0,0 +1,330 @@
+    """
+    Build R object.
+
+    Parameters
```
Aren't we using Google docstrings? Why this sudden change to NumPy style?

trossi replied:
Fixed! That was unintentional.

vnmabus commented on rdata/conversion/to_r.py:
```diff
@@ -0,0 +1,330 @@
+            object=False,
+            attributes=attributes is not None,
+            tag=tag is not None,
+            gp=int(gp),
```
Why the casting?

trossi replied:
Removed, should be unnecessary indeed.

vnmabus commented on rdata/conversion/to_r.py:
```diff
@@ -0,0 +1,330 @@
+    if convert_value is None:
+        convert_value = convert_to_r_object
+
+    if isinstance(data, dict):
```
How do we deal with empty dicts/lists?

trossi replied:
Good point.

But, empty dicts/lists convert to `RObjectType.VEC` type, so it remains unclear what kind of `RObjectType.LIST` object `build_r_list()` should return for empty input as this case is never encountered with current tests . So, I added `raise ValueError` to ensure non-empty inputs here. (If there is an example case for an empty `RObjectType.LIST` type, then we can replace raising the error with the corresponding logic here.)

However, there was a bug elsewhere in converting `{}` to an empty R named list, and it is fixed now.

I added also two new tests files `test_empty_list.rds` and `test_empty_named_list.rds` and tests reading and writing those. These files are created in R as follows:
```
$ Rscript -e "data = list(); print(data); print(typeof(data)); saveRDS(data, 'empty_list.rds')"
list()
[1] "list"

$ Rscript -e "data = list(a=1); data = data[-1]; print(data); print(typeof(data)); saveRDS(data, 'empty_named_list.rds')"
named list()
[1] "list"
```

For reference, R objects for these empty lists/dicts are not `RObjectType.LIST`, but instead empty list is:
```
RObjectType.VEC
  value:
```
and empty dictionary (or R's named list) is:
```
RObjectType.VEC
  value:
  attributes:
    RObjectType.LIST
      tag:
        RObjectType.SYM
          value:
            RObjectType.CHAR(gp=64)
              value:
                b'names'

      value:
        RObjectType.STR
          value:
        RObjectType.NILVALUE
          value:
            None
```

vnmabus replied:
I see. Well, as you said, we can raise for now until we know what we want to do in that case exactly.

vnmabus commented on rdata/conversion/to_r.py:
```diff
@@ -0,0 +1,330 @@
+    convert_to_r_object
+    """
+    if versions is None:
+        versions = RVersions(3, 262657, 197888)
```
Maybe put these as constants somewhere.

trossi replied:
Done and improved in general. I did some further testing and the format version (2 vs 3) seems to define the minimum R version, so I added:
```py
# Mapping from format version to minimum R version
R_MINIMUM_VERSIONS: Final[Mapping[int, int]] = MappingProxyType({
    2: 0x20300,
    3: 0x30500,
})
```
I tested R 3.1.3, 3.6.3, 4.4.0 and checked also through the test files in `tests/data/` in case some specific type written would trigger higher minimum R version, which wasn't the case in any of those files.

Hex format of the integers is also quite much nicer here (`197888 == 0x30500` i.e. R version 3.5.0).

trossi replied:
Related, I added hex representation to `RVersions.__str__()` 49f70a1.

vnmabus replied:
All changes here seem good.

vnmabus commented on rdata/conversion/to_r.py:
```diff
@@ -0,0 +1,330 @@
+        versions = RVersions(3, 262657, 197888)
+
+    minimum_version_with_encoding = 3
+    extra = RExtraInfo(encoding) if versions.format >= minimum_version_with_encoding \
```
I prefer parentheses rather than escaping the newline.

trossi replied:
Fixed.

vnmabus commented on rdata/conversion/to_r.py:
```diff
@@ -0,0 +1,330 @@
+        obj = convert_to_r_object(data, encoding=encoding)
+    else:
+        if not isinstance(data, dict):
+            msg = "For RDA file, data must be a dictionary."
```
Maybe we can include the actual type received in the message, to facilitate debugging.

trossi replied:
Good idea, done.

vnmabus commented on rdata/conversion/to_r.py:
```diff
@@ -0,0 +1,330 @@
+    attributes = None
+    tag = None
+
+    if data is None:
```
Note that we should drop Python 3.9, according to [SPEC0](https://scientific-python.org/specs/spec-0000/). Once that is done, this code (and equivalent code for parsing and conversion) could benefit from the [`match` statement](https://docs.python.org/3/reference/compound_stmts.html#match) added in Python 3.10.

trossi replied:
That would be nice. (I didn't change that here yet.)

vnmabus replied:
Yes, it was just an idea for later. Nothing to be done for now.

vnmabus commented on rdata/conversion/to_r.py:
```diff
@@ -0,0 +1,330 @@
+                                      encoding=encoding)
+
+    elif isinstance(data, np.ndarray):
+        if data.dtype.kind in ["O"]:
```
I don't understand this case.

trossi replied:
This is for the case `rdata/tests/data/test_na_string.rda` that reads in as `{'test_na_string': array([None], dtype=object)}`. This if-block can handle only this particular case (note `assert data.size == 1` and `assert data[0] is None` here). I added a comment explaining this and added `NotImplementedError` for other cases than this special case.

vnmabus replied:
I see. Thanks for adding clarification.

vnmabus commented on rdata/conversion/to_r.py:
```diff
@@ -0,0 +1,330 @@
+        elif encoding == "CP1252":
+            # Note!
+            # CP1252 and Latin1 are not the same.
+            # Does CharFlags.LATIN1 mean actually CP1252
```
I am not sure, but I do not think it matters much in practice. You can keep the note here for future reference.

trossi replied:
:+1:

vnmabus commented on rdata/parser/_parser.py:
```diff
@@ -351,6 +357,26 @@ class RObject:
     tag: RObject | None = None
     referenced_object: RObject | None = None
 
+    def __eq__(self, other: object) -> bool:
```
Why do we need to compare R objects?

trossi replied:
This is used in `assert r_data == new_r_data` in `test_convert_to_r()` testing R-to-Python-to-R round trip. This could be converted to a separate function instead though. What would you suggest?

vnmabus replied:
Now I get it. It is fine, I was just curious about why we needed it.

vnmabus commented on rdata/tests/test_write.py:
```diff
@@ -0,0 +1,162 @@
+    """Test unparsing RData object to a file."""
+    with (TESTDATA_PATH / fname).open("rb") as f:
+        data = decompress_data(f.read())
+        rds = data[:2] != b"RD"
```
Maybe we should create functions for peeking the file format, instead of doing it twice in both tests. If they are useful enough we could even consider making them public in the future.

trossi replied:
I added a helper function using `FileTypes`, `RdataFormats` etc from `_parser.py`. I think it would be useful to refactor these type and format checking so that they are convenient both for parsing and unparsing (related to the comment about `Literal` above). Maybe such refactoring would be better in a separate PR to avoid too many changes on parsing side in this PR?

vnmabus replied:
Yes, I agree with postponing that refactor.

vnmabus commented on rdata/tests/test_write.py:
```diff
@@ -0,0 +1,162 @@
+        if fname in [
+            "test_encodings.rda",     # encoding not kept in Python
+            "test_encodings_v3.rda",  # encoding not kept in Python
+            "test_list_attrs.rda",    # attributes not kept in Python
```
Hum, do you happen to know what these attributes are for? Maybe we should try to keep them somehow.

trossi replied:
File `test_list_attrs.rda` has object
```
RObjectType.LIST
  value:
    RObjectType.STR
      value:
        RObjectType.CHAR(gp=64)
          value:
            b'list'
    RObjectType.LIST
      value:
        RObjectType.REAL
          value:
            [5.]
        RObjectType.NILVALUE
          value:
            None
  attributes:
    RObjectType.LIST
      tag:
        RObjectType.SYM
          value:
            RObjectType.CHAR(gp=64)
              value:
                b'my_attr'

      value:
        RObjectType.STR
          value:
            RObjectType.CHAR(gp=64)
              value:
                b'attr_value'
        RObjectType.NILVALUE
          value:
            None
```
but after conversion to Python it is just `[array(['list'], dtype='<U4'), array([5.])]` without `my_attr`.

File `test_file.rds` has even more complex object which is largely ignored in conversion to Python (but there is also warning `Missing constructor for R class "file"`).

vnmabus replied:
I see. I am still not sure if we should preserve the list attributes. I think we can just wait until we have a use case for them.
For the files and so I did not care too much. There was no clear analog in Python, so I just wanted to be able to open datasets containing them without losing much information. Thus, it does not surprise me that I did not bother with class "file".

vnmabus commented on rdata/tests/test_write.py:
```diff
@@ -0,0 +1,162 @@
+
+        encoding = r_data.extra.encoding
+        if encoding is None:
+            encoding = "CP1252" if "win" in fname else "UTF-8"
```
What happens if we just use "UTF-8" everywhere? Are we sure that is the way R does it?

trossi replied:
The reason for this line is that the v2 test files have no encoding information in header, so we just assume that files with "win" in the name would be with CP1252 (hacky). In terms of functionality, it would be ok to use UTF-8 here always, but then this test of R-to-Python-to-R round trip would fail for those cases, because the created RData object would have differently encoded strings than the original RData object.

vnmabus replied:
But do we use the characters that differ between these two encodings? IIRC they were almost identical, except for some esoteric glyphs.

trossi replied:
UTF-8 and Latin-1 (or its variants) differ for all non-ascii characters. For example 'ñ' or 'ä' are quite common (and included in some test files) and they are different in UTF-8 and Latin-1.

vnmabus replied:
Sorry, in my mind I mixed this discussion with the CP1252 vs Latin-1 one above. Of course UTF-8 and CP1252 are very different.

Given that this hack is in a test, and that I might try to improve testing later, I think it is fair to accept this as is.

vnmabus commented on rdata/tests/test_write.py:
```diff
@@ -0,0 +1,162 @@
+
+
+@contextmanager
+def no_error() -> Generator[Any, Any, Any]:
```
Isn't this [`contextlib.nullcontext`](https://docs.python.org/3.7/library/contextlib.html#contextlib.nullcontext)?

trossi replied:
True, changed to `nullcontext`.

vnmabus commented on rdata/tests/test_write.py:
```diff
@@ -0,0 +1,162 @@
+@pytest.mark.parametrize("rds", [True, False])
+def test_write_file(compression: str, fmt: str, rds: bool) -> None:  # noqa: FBT001
+    """Test writing RData object to a real file with compression."""
+    expectation = no_error()
```
Maybe you can specify here the type as [`AbstractContextManager`](https://docs.python.org/3/library/contextlib.html#contextlib.AbstractContextManager), so that it is not necessary to use `# type: ignore [assignment]` later.

trossi replied:
Done, thanks! That works indeed.

vnmabus commented on rdata/unparser/__init__.py:
```diff
@@ -0,0 +1,119 @@
+        Compression (gzip, bzip2, xz, or none)
+    """
+    if compression == "none":
+        from builtins import open  # noqa: UP029
```
Maybe it is better to use a name different from `open`, to prevent confusion.

trossi replied:
Renamed to `open_with_compression`.

vnmabus commented on rdata/unparser/__init__.py:
```diff
@@ -0,0 +1,119 @@
+    compression:
+        Compression (gzip, bzip2, xz, or none)
+    """
+    if compression == "none":
```
I would rather use `None` instead of a string.

trossi replied:
Done.

vnmabus commented on rdata/unparser/__init__.py:
```diff
@@ -0,0 +1,119 @@
+        from gzip import open  # type: ignore [no-redef]
+    elif compression == "xz":
+        from lzma import open  # type: ignore [no-redef]
+    else:
```
I would remove this case, so that `None` means "no compression".

trossi replied:
Done.

vnmabus commented on rdata/unparser/_ascii.py:
```diff
@@ -0,0 +1,78 @@
+                line = "NA" if value is None or np.ma.is_masked(value) else str(value)  # type: ignore [no-untyped-call]
+
+            elif np.issubdtype(array.dtype, np.floating):
+                line = str(value)
```
Does this work for NaN and infs?

trossi replied:
Good point. It almost did, writing `nan` and `inf`, but it's now fixed to write `NaN` and `Inf` instead. I added also two new test files `rdata/tests/data/test_ascii_nan_inf.rds` and `rdata/tests/data/test_nan_inf.rds` and corresponding tests for parsing and unparsing.

These new test files are:
```
Rscript -e "data = c(0, NaN, +Inf, -Inf); saveRDS(data, 'ascii_nan_inf.rds', compress=FALSE, ascii=TRUE)"
Rscript -e "data = c(0, NaN, +Inf, -Inf); saveRDS(data, 'nan_inf.rds')"
```

vnmabus replied:
Good! These values are more common than what it would seem.

vnmabus commented on rdata/unparser/_unparser.py:
```diff
@@ -0,0 +1,170 @@
+                f"{info.tag:1b}"
+                f"{info.attributes:1b}"
+                f"{info.object:1b}"
+                )
```
Small formatting nitpick (so that everything is aligned and parentheses match):
```suggestion
        bits = (
            f"{0:4b}"
            f"{info.gp:16b}"
            f"{0:1b}"
            f"{info.tag:1b}"
            f"{info.attributes:1b}"
            f"{info.object:1b}"
        )
```

trossi replied:
Fixed.

vnmabus commented on rdata/unparser/_unparser.py:
```diff
@@ -0,0 +1,170 @@
+    bits += f"{info.type.value:8b}"
+    bits = bits.replace(" ", "0")
+    assert len(bits) == 32  # noqa: PLR2004
+    return np.packbits([int(b) for b in bits]).view(">i4").astype("=i4")[0]  # type: ignore [no-any-return]
```
Can we reduce the trickery in this line using the `bitorder` argument of `np.packbits`? Or at least use a comment. It is difficult to understand.

trossi replied:
Changed this to a much simpler way: `int(f"0b{bits}", base=2)`.

vnmabus replied:
Much much nicer.

vnmabus commented on rdata/unparser/_unparser.py:
```diff
@@ -0,0 +1,170 @@
+    return np.packbits([int(b) for b in bits]).view(">i4").astype("=i4")[0]  # type: ignore [no-any-return]
+
+
+class Unparser(abc.ABC):
```
I honestly think that we should use `Protocol` instead of `ABC` almost everywhere, as it allows duck-typing and has all the advantages of the latter (plus, it is composable with other protocols).

I don't see the advantage of plain `ABC`s for the majority of cases (AFAIK it is only useful for *preventing* duck-typing, which is usually undesirable).

trossi replied:
Good point. I'd leave this to be `Unparser(ABC)` though to be symmetric with `Parser(ABC)`. Later they both could be refactored to `Protocol` in a symmetric way.

vnmabus replied:
I agree.

vnmabus commented on rdata/unparser/_unparser.py:
```diff
@@ -0,0 +1,170 @@
+            RObjectType.REAL,
+            RObjectType.CPLX,
+        }:
+            self.unparse_array(value)
```
Should we consider ND arrays? IIRC we just needed to add some attributes, and it may be very useful for some users.

trossi replied:
ND arrays work already. The original ND array is flattened to 1D and attributes are added in `convert_to_r_object()`. Then this unparser line sees only the flattened array.

vnmabus replied:
I see. Thanks for the clarification.

vnmabus reviewed (changes requested):
Sorry for taking so long to review it. You have put a lot of work and attention to detail here!

It is mostly good, except for the docstring format and some comments I made.

trossi reviewed (commented):
@vnmabus Thank you for the review! Very useful comments that improved the code. I have addressed the comments and included a few extra tests too.

vnmabus commented on rdata/tests/test_write.py:
```diff
@@ -0,0 +1,213 @@
+    with (TESTDATA_PATH / fname).open("rb") as f:
+        data = decompress_data(f.read())
+        file_type, file_format = parse_file_type_and_format(data)
+        r_data = rdata.parser.parse_data(data, expand_altrep=False)
```
Why `expand_altrep=False`?

trossi replied:
I kept altrep types in these roundtrip tests as I feel that having automatic de-expand in unparsing could be problematic in general (for example, would sequences like `[0, 1, 2]` be written as a such vector or mapped to an altrep R-type similar to `range(3)`).

vnmabus commented on rdata/unparser/__init__.py:
```diff
@@ -0,0 +1,125 @@
+        compression: Compression.
+    """
+    if compression is None:
+        open_with_compression = open
```
I think it is also possible to type this line (with a `Callable`) instead of using `# type: ignore [assignment]` later.

trossi replied:
I didn't manage to get a typehint work here. I keep getting `Incompatible types in assignment (expression has type overloaded function, variable has type "Callable[...]")` and similar for `open_with_compression` with different typehints. One way to make mypy happy here would be to duplicate `with open(...): unparse_fileobj(...)` under each if-block, but such code duplication wouldn't be that nice.

vnmabus replied:
Well, I will merge as is, and I will try to change it latter.

vnmabus commented on rdata/conversion/to_r.py:
```diff
@@ -0,0 +1,339 @@
+        r_type = RObjectType.CHAR
+        if all(chr(byte) in string.printable for byte in data):
+            gp = CharFlags.ASCII
+        elif encoding == "UTF-8":
```
If arbitrary encodings are not supported, maybe `encoding` should be a literal too. This would also prevent users to use a different spelling (e.g. `"utf-8"`). We should also consider which spelling do we want to support (I think in Python `"utf-8"` is more common).

trossi replied:
I updated type to Literal and changed to lower case. (Upper case was used earlier as that's how R writes the encoding to the file.) This changed required some extra type ignores in tests as `RExtraInfo` can contain any encoding (`str` type).

In principle we could allow any encoding in writing, but that would require more testing on which encodings R can read in, so restricting to UTF-8 and CP1252 for now is safer in that sense.

In general, I think it would be nice to have something like
```python
class Encoding(enum.Enum):
    """Encoding for strings."""

    UTF8 = "UTF8"
    CP1252 = "CP1252"

    def __eq__(self, other: Any) -> bool:
        if isinstance(other, Encoding):
            if other is self:
                return True
        if isinstance(other, str):
            if other.upper().replace("-", "") == self.value:
                return True
        return False
```

Then user can provide `Encoding.UTF8`, `'utf8'`, `'UTF-8'` and they would all be equal to `Encoding.UTF8`.  But this could be left for later refactor - similar enums could be nice for file formats and compression too.

vnmabus replied:
That seems good.

vnmabus reviewed (changes requested):
It is almost ready. Some small comments and questions.

trossi reviewed (commented):
@vnmabus Thank you for the review! I replied to comments and fixed also a NumPy 2.0 compatibility issue.

vnmabus reviewed (approved):
LGTM. I think this can be merged now. If we have future improvements, they can be done in a different PR.

About the JOSS paper, there is an open discussion in https://github.com/pyOpenSci/software-submission/issues/144 , because usually in order to be fast-tracked, there should be no changes between pyOpenSci acceptance and JOSS submission, and for this package we have made several changes in the middle. My hope is that at least we could submit it using the normal submission procedure (without fast-tracking), but we need to wait for answers as this situation is highly unusual.

trossi commented:
@vnmabus Thank you for the thorough review and the update regarding JOSS!
